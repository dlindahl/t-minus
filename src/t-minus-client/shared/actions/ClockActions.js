import ClockActions from '../actionTypes/ClockActionTypes';
import emptyObj from 'empty/object';
import formatDuration from '../util/formatDuration';
import partial from 'lodash/partial';
import present from 'present';

export const IntervalValue = 50;
let interval = null;

function calculateDuration(mode, elapsedTime, maxDuration) {
  switch(mode) {
    case ClockActions.CLOCK_MODE_TIMER:
      const timeRemaining = maxDuration - elapsedTime;
      const percentComplete = Math.min(elapsedTime / maxDuration, 1);
      return [formatDuration(timeRemaining), timeRemaining, percentComplete];
    default:
      return [formatDuration(elapsedTime), null, 1]
  }
}

function determineTimerElapsed(mode, elapsedTime, maxDuration) {
  switch(mode) {
    case ClockActions.CLOCK_MODE_TIMER:
      return elapsedTime > maxDuration;
    default:
      return false;
  }
}

export function clockTick(dispatch, getState) {
  const { clock } = getState();
  if(!clock.running) {
    return;
  }
  const currentTick = present();
  const delta = currentTick - clock.lastTick;
  const elapsedTime = clock.elapsedTime + delta;
  const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, elapsedTime, clock.timerValue);
  const hasTimerElapsed = determineTimerElapsed(clock.mode, elapsedTime, clock.timerValue);
  dispatch({
    type: ClockActions.CLOCK_TICK,
    payload: {
      currentTick,
      delta,
      duration,
      elapsedTime,
      hasTimerElapsed,
      mode: clock.mode,
      percentComplete,
      timeRemaining
    }
  });
}

export function pauseClock() {
  return {
    type: ClockActions.CLOCK_PAUSED,
    payload: emptyObj
  };
}

export function resetClock() {
  return (dispatch, getState) => {
    const { clock } = getState();
    clearInterval(interval);
    interval = null;
    const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, 0, clock.timerValue);
    return dispatch({
      type: ClockActions.CLOCK_RESET,
      payload: {
        duration,
        timeRemaining,
        timerValue: clock.timerValue,
        percentComplete
      }
    });
  }
}

export function startClock() {
  return (dispatch, getState) => {
    const { clock } = getState();
    const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, clock.elapsedTime, clock.timerDuration);
    dispatch({
      type: ClockActions.CLOCK_STARTED,
      payload: {
        duration,
        percentComplete,
        startedAt: present(),
        timeRemaining
      }
    });
    interval = setInterval(partial(clockTick, dispatch, getState), IntervalValue);
  };
}

export function changeTimerValue(timerValue) {
  return (dispatch, getState) => {
    const { clock } = getState();
    const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, 0, timerValue);
    return dispatch({
      type: ClockActions.CLOCK_TIMER_CHANGED,
      payload: { duration, percentComplete, timeRemaining, timerValue }
    });
  }
}

export function stopwatchMode() {
  const mode = ClockActions.CLOCK_MODE_STOPWATCH;
  const [duration, timeRemaining, percentComplete] = calculateDuration(mode, 0);
  return {
    type: ClockActions.CLOCK_MODE_CHANGED,
    payload: {
      duration,
      mode,
      percentComplete,
      timeRemaining
    }
  }
}

export function timerMode() {
  return (dispatch, getState) => {
    const { clock } = getState();
    const mode = ClockActions.CLOCK_MODE_TIMER;
    const [duration, timeRemaining, percentComplete] = calculateDuration(mode, 0, clock.timerValue);
    return dispatch({
      type: ClockActions.CLOCK_MODE_CHANGED,
      payload: {
        duration,
        mode,
        percentComplete,
        timeRemaining
      }
    });
  }
}

export function toggleClock() {
  return (dispatch, getState) => {
    const clock = getState().clock
    if(clock.startedAt == null) {
      startClock()(dispatch, getState);
    } else if(clock.running) {
      dispatch(pauseClock());
    } else {
      dispatch(unpauseClock());
    }
  };
}

export function unpauseClock() {
  return {
    type: ClockActions.CLOCK_UNPAUSED,
    payload: {}
  };
}
