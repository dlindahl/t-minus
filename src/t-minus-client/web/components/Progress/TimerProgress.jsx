import { assign } from 'lodash'
import ClockActions from '../../../shared/actionTypes/ClockActionTypes'
import { PropTypes } from 'react'
import { connect } from 'react-redux'
import withSeverity from '../../decorators/withSeverity'

const DefaultHeight = 50
const MaxWidthPercentage = 100

const baseStyles = {
  empty: {
    height: DefaultHeight
  },
  progress: {
    height: DefaultHeight,
    whiteSpace: 'nowrap'
  }
}

function getState (state) {
  return {
    mode: state.clock.mode,
    percentComplete: state.clock.percentComplete,
    severity: state.display.severity
  }
}

const TimerProgress = (props) => {
  let progressBackgroundColor = props.backgroundSecondaryColor
  if (props.mode !== ClockActions.CLOCK_MODE_TIMER) {
    progressBackgroundColor = props.backgroundPrimaryColor
  }
  const containerStyle = {
    background: props.backgroundPrimaryColor
  }
  const progressStyle =
    assign(
      {},
      baseStyles.progress,
      {
        background: progressBackgroundColor,
        width: `${props.percentComplete * MaxWidthPercentage}%`
      }
    )
  return (
    <div style={containerStyle}>
      {props.children}
      <div style={progressStyle}/>
    </div>
  )
}

TimerProgress.propTypes = {
  backgroundPrimaryColor: PropTypes.string,
  backgroundSecondaryColor: PropTypes.string,
  children: PropTypes.node,
  mode: PropTypes.string,
  percentComplete: PropTypes.number
}
TimerProgress.defaultProps = {
  percentComplete: 0.5
}

export default withSeverity(connect(getState)(TimerProgress))
