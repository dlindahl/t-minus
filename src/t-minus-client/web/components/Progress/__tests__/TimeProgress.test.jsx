import ClockActions from '../../../../shared/actionTypes/ClockActionTypes';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import TimerProgress from '../TimerProgress';
import TimerSeverity from '../../../../shared/actionTypes/TimerSeverityTypes';

const states = {
  [ClockActions.CLOCK_MODE_TIMER]: {
    clock: {
      mode: ClockActions.CLOCK_MODE_TIMER,
      percentComplete: .5
    },
    display: {
      severity: TimerSeverity.TIMER_SEVERITY_CALM
    }
  },
  [ClockActions.CLOCK_MODE_STOPWATCH]: {
    clock: {
      mode: ClockActions.CLOCK_MODE_STOPWATCH,
      percentComplete: .5
    },
    display: {
      severity: TimerSeverity.TIMER_SEVERITY_CALM
    }
  }
};
const store = configureStore();

describe('TimeProgress', () => {
  it('renders a timer when in timer moder', () => {
    const component = renderer.create(
      <TimerProgress store={store(states[ClockActions.CLOCK_MODE_TIMER])}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a stopwatch when in stopwatch moder', () => {
    const component = renderer.create(
      <TimerProgress store={store(states[ClockActions.CLOCK_MODE_STOPWATCH])}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
