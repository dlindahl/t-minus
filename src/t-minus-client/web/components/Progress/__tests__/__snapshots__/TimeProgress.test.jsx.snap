exports[`TimeProgress renders a stopwatch when in stopwatch moder 1`] = `
<div
  style={
    Object {
      "background": "#bdc3c7",
      "height": 50,
    }
  } />
`;

exports[`TimeProgress renders a timer when in timer moder 1`] = `
<div
  style={
    Object {
      "background": "#bdc3c7",
    }
  }>
  <div
    style={
      Object {
        "background": "#95a5a6",
        "height": 50,
        "whiteSpace": "nowrap",
        "width": "50%",
      }
    } />
</div>
`;
